### ISSUE TITLE
Increasing performance on Raspberry pi

### ISSUE THREAD

**Issue by pourfard** [alphacep/vosk-api#331]
Hi,

I have tested vosk api on raspberry pi 3 B. It takes about 90 seconds to recognize a 60 seconds audio file. During processing it takes about 25% of CPU ( I think it runs only on one core).

Is there any way to use more cores and speed up speech recognition on raspberry pi? The only way that I'm thinking about is overclocking. I was wondering if exists an environment variable for vosk or kaldi to define number of processing threads?

**Comment by nshmyrev** [comment]
There are no easy ways unfortunately, you can only restrict the vocabulary or significantly reduce beams (edit max-active and beam in model/conf/model.conf). Master issue is 

Main issue for this problem with some more details:

https://github.com/alphacep/vosk-api/issues/326

**Comment by nshmyrev** [comment]
On 3B+ it runs in realtime, so it takes about 60 seconds to decode 60 second file. We might tune the system a bit later since fast execution is certainly in our plans.

As for multicore computation, it requires coding but in general possible.

**Comment by pourfard** [comment]
Thank you so much,

But title of the main thread says it is sub-realtime on rpi 4. How can it
be possible?

Rpi 3 b: 1.2 ghz (1.5 with overclocking)
Rpi 3 b+: 1.4 ghz
Rpi 4: 1.5 ghz

**Comment by nshmyrev** [comment]
It is overall much slower than possible due to suboptimal implementation of various components.

**Comment by pourfard** [comment]
Have you ever tried on 64 bit OS? I read in an issue that someone have
compiled and used vosk on 64 bit OS. Is it possible to increase perpormance
in such a way?

**Comment by nshmyrev** [comment]
I didn't try myself. 64-bit build is available in releases, you can try if you have the OS.

No guarantees it will be faster though.

**Comment by pourfard** [comment]
I'm so sorry,

The sample rate of the file that I was testing was 22050Hz but the model was 16000. I changed the sample rate and every thing is ok. Using both file and microphone is realtime.

The only thing that I need now is how to introduce new words to the model. Based on the documentation I followed [this post](https://chrisearch.wordpress.com/2017/03/11/speech-recognition-using-kaldi-extending-and-using-the-aspire-model/) but it is based on a pretrained big model from kaldi (which does not exists for our language). Can we add a new word to [vosk-model-small-fa-0.4 ](https://alphacephei.com/vosk/models/vosk-model-small-fa-0.4.zip) or the only way for expanding vocabulary is giving you a list of words as you said in this issue #321?
